[`@react-native/metro-config`] Added inference of `watchFolders` from NPM/Yarn workspaces

kraenhansen opened the pull request:
## Summary:

When serving the Metro server from a mono-repo leveraging NPM or Yarn workspaces, the developer need to manually configure Metro to include the workspace root in the `watchFolders` (or `projectRoot` as per the suggestion in the Metro config docs).

I'm aware the [team might already be working on a solution](https://reactnative.dev/blog/2023/12/06/0.73-debugging-improvements-stable-symlinks#monorepo-workarounds) for this, but I wanted to suggest this in case you hadn't started work on it or hadn't considered this approach.

### Resolving the workspace root

I initially experimented with a solution using [@npmcli/arborist](https://www.npmjs.com/package/@npmcli/arborist) to resolve the workspace root, but I stopped and implemented the `getWorkspaceRoot` function instead as soon as I realised the package only provided async APIs to load the "actual tree".

This PR suggests adding a `getWorkspaceRoot` function to resolve the workspace root, which supports NPM and Yarn workspaces. It use the same dependency to resolve globs (`fast-glob`) which [may occur in Yarn workspaces](https://classic.yarnpkg.com/lang/en/docs/workspaces/#toc-tips-tricks). I've double-checked that neither NPM nor Yarn supports workspaces outside of descendent folders, i.e. the root will always be in a parent directory of a package.

I've considered using existing packages to find the workspace root, but I wanted a solution that would be robust to malformed `package.json` files and thought the algorithm was simple enough to inline here. If the reviewer thinks this is not important enough, I can update the PR to use the [find-yarn-workspace-root](https://www.npmjs.com/package/find-yarn-workspace-root), which seem to be a fairly safe dependency to take on (with 2 mill. weekly downloads).

## Changelog:

[GENERAL] [ADDED] - Add automatic resolution of NPM / Yarn workspaces root as `watchFolders` in the `@react-native/metro-config` package.

## Test Plan:

I've tested this in an app locally.
The package doesn't have tests setup at the moment, but I'd be happy to contribute some if that's needed / desired by the reviewer.

robhogan commented:
Hey @kraenhansen  - love this in general as a pragmatic (but careful) default heuristic, thanks for the contribution.

> The Metro config [suggests using the projectRoot](https://metrobundler.dev/docs/configuration/#projectroot) when configuring for a mono-repo: [...] I wasn't able to find a way to configure [the app entry point] back to using the original `projectRoot` passed by the user and found it would be simpler to just use the `watchFolders` for this use-case. I'd love some feedback on this decision.

I agree (see https://github.com/facebook/metro/issues/1016), - both approaches are valid and I don't think the advice in the Metro docs is the path of least resistance for typical single app Metro instances (the tradeoffs are different for running Metro standalone, serving multiple apps), but we haven't had chance to update those docs.

> thought the algorithm was simple enough to inline here

👍

> NPM / Yarn workspaces root

My main hesitation with this PR is introducing a split between support for Yarn/NPM workspaces vs pnpm workspaces - was this something you looked into? One possible complication is that if pnpm also supports nested roots we'd need to add the complication of Yaml parsing, but glob matching should be the same. I wouldn't say pnpm support is a blocker though.

robhogan commented on packages/metro-config/index.js:
```diff
@@ -36,6 +39,53 @@ const INTERNAL_CALLSITES_REGEX = new RegExp(
+ */
+function getWorkspaceRoot(projectRoot /*: string */, candidatePath /*: string */ = projectRoot) /*: string | void */ {
+  const packageJsonPath = path.resolve(candidatePath, 'package.json');
+  if (fs.existsSync(packageJsonPath)) {
```
Since we're reading outside the project there's a reasonable chance of a permissions error - I'd be inclined to treat inaccessible files as effectively non-existent. WDYT?

```suggestion
  if (fs.accessSync(packageJsonPath, fs.constants.R_OK)) {
```

kraenhansen replied:
It seems `fs.accessSync` doesn't return a boolean like `existsSync` did. I believe I can simply remove this guard and instead make the `catch` below ignore an error with `code === 'ENOENT'`.

robhogan commented on packages/metro-config/index.js:
```diff
@@ -36,6 +39,53 @@ const INTERNAL_CALLSITES_REGEX = new RegExp(
+        }
+      }
+    } catch (err) {
+      console.warn(`Failed reading or parsing ${packageJsonPath}:`, err);
```
Hmm.. this is awkward, because if there is an error here for whatever reason (less chance if we check for permissions above), even if the user explicitly specifies `watchFolders`, they can't (in this design) avoid running this auto-detection, generating warnings that aren't relevant to them.

Ideally, this detection would only run if the user hasn't specified `watchFolders` themselves, but to do that I think we'd need to move it to the `@react-native/community-cli-plugin` - and tbh that might be a better place for it than amongst these static defaults. (WDYT @huntie ?)

huntie replied:
I think that sounds fair 👍🏻

[Expo does perform similar detection as part of its config defaults](https://github.com/expo/expo/blob/2a96b86e780a2cc6c4cf4d5b38849f108f0517b0/packages/%40expo/metro-config/src/ExpoMetroConfig.ts#L154), however that is (conceptually) downstream from this package. I'd rather locate config that is dependent on the _embedder_ (Community CLI, Expo), particularly if dynamic, outside this file. So yeah, into `@react-native/community-cli-plugin`.

kraenhansen commented on packages/metro-config/index.js:
```diff
@@ -36,6 +39,53 @@ const INTERNAL_CALLSITES_REGEX = new RegExp(
+      if (Array.isArray(workspaces)) {
+        // If one of the workspaces match the project root, this is the workspace root
+        // Note: While NPM workspaces doesn't currently support globs, Yarn does.
+        const matches = fastGlob.sync(workspaces, {
```
I agree, this could be refactored to be less IO intensive. I'll update this to use `micromatch` (which is the underlying dependency of fast-glob)

kraenhansen commented on packages/metro-config/index.js:
```diff
@@ -36,6 +39,53 @@ const INTERNAL_CALLSITES_REGEX = new RegExp(
+ * @param {string | undefined} candidatePath Current path to search from
+ * @returns Path of a workspace root or `undefined`
+ */
+function getWorkspaceRoot(projectRoot /*: string */, candidatePath /*: string */ = projectRoot) /*: string | void */ {
```
Woups - I committed this too quickly, the `candidatePath` has a default value, so it should be `string`.

huntie commented on packages/metro-config/index.js:
```diff
@@ -36,6 +39,53 @@ const INTERNAL_CALLSITES_REGEX = new RegExp(
+ * - which has a `workspaces` array of strings
+ * - which (possibly via a glob) includes the project root
+ * @param {string} projectRoot Project root to find a workspace root for
+ * @param {string | undefined} candidatePath Current path to search from
```
Nit: As we have parameter types defined in code (Flow), it's redundant to repeat them in JSDoc (although we don't currently perform any processing on this — but let's simplify maintenance).

```suggestion
 * @param projectRoot Project root to find a workspace root for
 * @param candidatePath Current path to search from
```

IMO we could **omit `@param/@returns` entirely**, since `projectRoot` is self-documenting and `candidatePath` is explained in the first line.

kraenhansen replied:
I've removed the `@param` and `@returns` 👍

kraenhansen commented:
I moved the implementation into the `community-cli-plugin` package and updated the changelog entry. If you're happy with the approach, I could continue by adding support for pnpm workspaces too (taking on a dependency of a yaml parser) or we could merge this and I'd be happy to do a follow-up PR for that.

@robhogan / @huntie: Do you have any thoughts on the use of `watchFolders` vs `projectRoot` as I've mentioned in the description of the PR.

robhogan commented:
Thanks again @kraenhansen! I'm happy for pnpm to be a follow-up.

Re `watchFolders` vs `projectRoot`, I mentioned that a bit above (more in the linked issue), but tldr I agree that customising `watchFolders` is the right approach for the RN template default.

kraenhansen commented on packages/community-cli-plugin/src/utils/loadMetroConfig.js:
```diff
@@ -71,6 +77,12 @@ function getOverrideConfig(
     },
   };
+
+  if (config.watchFolders.length === 0) {
+    overrides.watchFolders = getWatchFolders(ctx.root);
+  }
```
The only problem I see with this approach of providing the default as an override after loading the user-land config, is that the developer has to provide at least one element in the `watchFolders` to disable this default.

kraenhansen replied:
Another alternative would be to update the `metro-config` package to not provide an empty array for `watchFolders` and match on `undefined` instead of an empty array here. But .. that might be slightly breaking to developers that use the default `watchFolders` array by pushing their values into it 🤔

kraenhansen commented on packages/community-cli-plugin/src/utils/loadMetroConfig.js:
```diff
@@ -71,6 +72,17 @@ function getOverrideConfig(
+
+  // Applying the heuristic of appending workspace root as watch folder,
+  // only if no other watch folder (beside the project root) has been given.
+  if (!config.watchFolders.some(folder => folder !== ctx.root)) {
```
The only potential problem I see with this approach of providing the default as an override after loading the user-land config, is that the developer has to provide at least one element (other than the project root which is automatically injected) in the watchFolders to disable this default heuristic.

huntie replied:
Yeah it's tricky 😅.

I think let's bypass this logic if `config.watchFolders != null`. i.e. _**any value set by the user is an opt-out**_. We'd also want to note this [in the docs](https://reactnative.dev/docs/metro) — under a header "Workspaces detection" or similar (which could also briefly describe this feature to users 🙂).

kraenhansen replied:
I did some experiments passing the workspace root via the second parameter to `metro-config`'s `loadConfig` function.

It turns out that even if we omitted `watchFolders` from the config here: https://github.com/facebook/react-native/blob/f1a7f08feb2a63926e91d37b5d85130c7b74ee31/packages/metro-config/index.js#L85 it would still propagate as `[]` to the `community-cli-plugin` because the `metro-config` package as `watchFolders: []` here: https://github.com/facebook/metro/blob/main/packages/metro-config/src/defaults/index.js#L163 and we merge on top of that:
https://github.com/facebook/react-native/blob/f1a7f08feb2a63926e91d37b5d85130c7b74ee31/packages/metro-config/index.js#L91-L94

I tried explicitly setting the `watchFolders` to `undefined` in `@react-native/metro-config`, but `metro-config`'s `loadConfig` expects this to be iterable: https://github.com/facebook/metro/blob/main/packages/metro-config/src/loadConfig.js#L326 🤔 

The only way I can get it working is by deleting the `watchFolders` property on the object returned from `getDefaultValues` here: https://github.com/facebook/react-native/blob/f1a7f08feb2a63926e91d37b5d85130c7b74ee31/packages/metro-config/index.js#L92 while omitting the `watchFolders` from the `config` in `@react-native/metro-config` entirely.

I've pushed another branch to illustrate the changes we'll need: https://github.com/kraenhansen/react-native/compare/kh/metro-config-workspace-root...kraenhansen:react-native:kh/metro-config-workspace-root-defaults?expand=1

huntie replied:
> metro-config's loadConfig expects this to be iterable: [facebook/metro@main/packages/metro-config/src/loadConfig.js#L326](https://github.com/facebook/metro/blob/main/packages/metro-config/src/loadConfig.js?rgh-link-date=2024-01-02T16%3A12%3A59Z#L326)

Interesting! I think we should fix that in Metro as `watchFolders` can be `undefined` as indicated in `InputConfigT`.

---

Separately, and back to the problem at hand (forgive me if I've lost context on something obvious) — can we instead:

- Keep the default of `watchFolders = []`.
- Apply the heuristic if `watchFolders` is `[]` or `== null`, and do nothing otherwise.
    - What's relevant to the user is _have they set an override for `watchFolders` in their `metro.config.js`?_ (i.e. in the [template](https://github.com/facebook/react-native/blob/main/packages/react-native/template/metro.config.js#L9)). If not, then we can apply this :)
    - This would be [documented](https://reactnative.dev/docs/metro) as ~"**`watchFolders` behaviour and workspaces detection**: The default for `watchFolders` is `[]`. If you configure a value to an array containing anything other than this, workspaces detection will not be applied."

kraenhansen replied:
Okay. I misunderstood you then. I thought you wanted end-users to be able to use an empty array to indicate that they want to opt out of the workspace detection.

The [metro-config's loadConfig injects the project root as a watch folder](https://github.com/facebook/metro/blob/main/packages%2Fmetro-config%2Fsrc%2FloadConfig.js#L325), so from the cli plugin, it won't ever be empty, but instead I suggested to check if it contains any other value than the project root.

robhogan replied:
Hmm - a user may explicitly specify `watchFolders: []` and in that case we should *not* override it by applying auto-detection. A referential check *could* work (to differentiate explicit `[]` from the default `[]`), though the clearest API.

(IMO, this would be easier if `metro.config.js` exported a partial `InputConfigT`, as it used to, rather than a full `ConfigT` with defaults already merged and explicit intent lost, as it does post `@react-native/metro-config`. I don't fully recall why we did it that way @huntie? I think it was discussed and I'm just forgetting the trade-offs.)

huntie replied:
Hmm. That default config merge is applied inside `@react-native/metro-config`, so we have control. The partial config was our initial approach, pivoted away from here:

- https://github.com/facebook/react-native/pull/38092

Now that we've [documented Metro config in RN better](https://reactnative.dev/docs/metro#advanced-using-a-config-function), perhaps we could move forward with the following changes in `@react-native/metro-config`:
- Removing the default config merge.
- Removing `watchFolders: []` in these defaults.
- Artificially reading and assigning the defaults for `sourceExts` and `assetExts` (to fit user expectations as this is frequently extended from).

kraenhansen replied:
I finally found some time to get back to this:
- I've implemented a couple of tests, including a failing test for the case where a developer explicitly set their `watchFolder: []` in the `metro.config.js`.
- I've realised, that even if we fix the bug with `metro-config` not taking `watchFolders: undefined`, I would have to move out the injection of `configWithArgs.projectRoot` over to `community-cli-plugin` (to determine if the developer left out the value or explicitly set it to the empty array) .. but that could break someone relying on `metro-config` injecting this 🤔 Any ideas on how to get around that? Or if that would be okay?
I can push a PR to metro to illustrate the change, but basically I would move [the entire `overriddenConfig` part](https://github.com/facebook/metro/blob/18210320c0466499f7874abc26eac1fe60ec519c/packages/metro-config/src/loadConfig.js#L322-L334) into [the `community-cli-plugin`'s overrides](https://github.com/facebook/react-native/blob/07e8ae42bed71f54bbe0e786ccad88b2f14648a4/packages/community-cli-plugin/src/utils/loadMetroConfig.js#L32C10-L32C27).

kraenhansen commented:
I pushed support for an alternative way of declaring workspaces in Yarn.

kraenhansen commented:
I also got around to adding support for PNPM workspaces and I choose to add a dependency on `yaml`, since that was already used by `@react-native-community/cli-doctor` and in the yarn.lock ... now, I'll let the dust settle for you to properly review 😅

kraenhansen commented on packages/community-cli-plugin/src/utils/getWorkspaceRoot.js:
```diff
@@ -0,0 +1,84 @@
+ * - [PNPM workspaces](https://pnpm.io/workspaces)
+ */
+function getWorkspacePaths(packagePath /*: string */) /*: string[] */ {
+  const result /*: string[] */ = [];
```
This function is pushing into an array of paths to avoid giving a preference to neither NPM/Yarn nor PNPM if someone has both configured. Before it would do an early return as soon as it found a `workspaces` array in the `package.json`.

kraenhansen replied:
No valid use case, I just wanted the code to be agnostic, but I can look for the `pnpm-workspace.yaml` first 👍

kraenhansen commented on packages/community-cli-plugin/src/utils/getWorkspaceRoot.js:
```diff
@@ -0,0 +1,84 @@
+  const parentDir = path.dirname(candidatePath);
+  if (parentDir !== candidatePath) {
+    return getWorkspaceRoot(projectRoot, parentDir);
+  } else {
+    return null;
+  }
```
```suggestion
  }
  return null;
```

kraenhansen commented on packages/community-cli-plugin/src/utils/__tests__/getWorkspaceRoot-test.js:
```diff
@@ -0,0 +1,83 @@
+const { getWorkspaceRoot } = require('../getWorkspaceRoot');
+const fs = require('node:fs');
+const os = require('node:os');
+const path = require('node:path');
```
Done 👍

kraenhansen commented on packages/community-cli-plugin/src/utils/getWorkspaceRoot.js:
```diff
@@ -0,0 +1,84 @@
+    }
+  } catch (err) {
+    if (err.code !== 'ENOENT') {
+      console.warn(`Failed getting workspace root from ${packagePath}:`, err);
```
I noticed the `loadMetroConfig` file was already using `logger.debug`: https://github.com/facebook/react-native/blob/main/packages/community-cli-plugin/src/utils/loadMetroConfig.js#L93

kraenhansen commented:
The tests pass locally if I patch `metro-config` by deleting the injection of the project root (see [my comment above](https://github.com/facebook/react-native/pull/41967#discussion_r1475235748)) and instead return `configWithArgs` early.
The injection is still happening, but now it has been moved into the `community-cli-plugin`: 

https://github.com/facebook/metro/blob/18210320c0466499f7874abc26eac1fe60ec519c/packages/metro-config/src/loadConfig.js#L322-L334

kraenhansen commented on packages/community-cli-plugin/src/utils/loadMetroConfig.js:
```diff
@@ -107,10 +124,16 @@ This warning will be removed in future (https://github.com/facebook/metro/issues
-    cwd,
-    ...options,
-  });
+  const config = await loadConfig(
+    {
+      cwd,
+      ...options,
+    },
+    {
+      // Enables users to explicitly specify watchFolders
+      watchFolders: undefined,
+    },
+  );
```
This would need an update once https://github.com/facebook/metro/pull/1206 is merged, released and pulled in.

```suggestion
  const config = await loadConfig(
    {
      cwd,
      ...options,
    },
    {
      // Enables users to explicitly specify watchFolders
      watchFolders: undefined,
    },
    false
  );
```

kraenhansen commented:
Please keep this open, I'll rebase as I return from vacation🤞
